### PR TITLE
verify and solve #223 with unit tests and disabling shell mode

### PR DESCRIPTION
**Pull request by simskij**
solves #223

### CODE REVIEW

**Review by sed-i — APPROVED** [review]

File: lib/charms/prometheus_k8s/v0/prometheus_remote_write.py [hunk 1]
```diff
@@ -1104,8 +1104,6 @@ def _apply_label_matcher(self, expression, topology):
             return self._exec(args)
         except Exception as e:
             logger.debug('Applying the expression failed: "{}", falling back to the original', e)
```
Comment (Abuelodelanada) [inline lib/charms/prometheus_k8s/v0/prometheus_remote_write.py:1106]: Is it possible to catch a more specific exception?

**Review by rbarry82 — APPROVED** [review]

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 2]
```diff
@@ -2222,8 +2222,6 @@ def _apply_label_matcher(self, expression, topology):
             return self._exec(args)
         except Exception as e:
             logger.debug('Applying the expression failed: "{}", falling back to the original', e)
```
Comment (Abuelodelanada) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:2224]: Is it possible to catch a more specific exception?

File: lib/charms/prometheus_k8s/v0/prometheus_remote_write.py (hunk 1 above)
Comment (simskij) [reply lib/charms/prometheus_k8s/v0/prometheus_remote_write.py:1106]: good question! regardless of what happened, we want it to catch and return the original, so I think it's appropriate in this case.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 2 above)
Comment (simskij) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:2224]: see previous comment

File: lib/charms/prometheus_k8s/v0/prometheus_remote_write.py (hunk 1 above)
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus_remote_write.py:1106]: `_exec` can raise the combination of `subprocess.run` and `result.stdout.decode` which is quite a bit, and I guess it doesn't matter why it failed.

Comment (rbarry82) [reply lib/charms/prometheus_k8s/v0/prometheus_remote_write.py:1106]: `decode` is really unlikely to fail unless the LOCALE has been changed inside a pod somehow.

```python
try:
  subprocess.run(...
except subprocess.CalledProcessError as e:
  # do something
```

Is probably a sufficient guard. I'm more surprised that the linters aren't yelling about the bare `Exception`

Comment (simskij) [reply lib/charms/prometheus_k8s/v0/prometheus_remote_write.py:1106]:
```
# noinspection PyBroadException
```

they would have, but I've suppressed them for that part as it, as I said, is intentional. 👍🏽

**Pull request merged**